Ticket #35831 Include ModelForm and its Meta options in reference docs, follow up on PR #18699 (#19758)
django/django: docs/ref/forms/models.txt

===========
Model Forms
===========

.. module:: django.forms.models
    :synopsis: ModelForm API reference for inner ``Meta`` class and factory
        functions

.. class:: ModelForm

.. _ref-modelform:

.. currentmodule:: django.forms

:class:`~django.forms.ModelForm` API reference. For introductory material about

> ModelForms, see the :doc:`/topics/forms/modelforms` topic guide.

Reviewer: ```suggestion
using a ``ModelForm``, see the :doc:`/topics/forms/modelforms` topic guide.
```

.. _ref-modelform-meta-options:

ModelForm ``Meta`` API
======================

.. class:: ModelFormOptions

> The ``_meta`` API is used to build forms that reflect a Django model. The API
> is accessible through the ``_meta`` attribute of each model form, which is an
> instance of a ``django.forms.models.ModelFormOptions`` object.
> The exact structure of the form produced can be controlled by setting metadata
> options as attributes of an inner ``Meta`` class.
> 
> Here is an example using some sort of Book model::

Reviewer: ```suggestion
The ``_meta`` API is used to build forms that reflect a Django model. It is
accessible through the ``_meta`` attribute of each model form, and is an
``django.forms.models.ModelFormOptions`` instance.

The structure of the generated form can be customized by defining metadata
options as attributes of an inner ``Meta`` class. For example::
```
(perhaps)

    from django.forms import ModelForm
    from myapp.models import Book

    class BookForm(ModelForm):
        class Meta:
            model = Book
            fields = ["title", "author"]
            help_texts = {
                "title": "The title of the book",
                "author": "The author of the book",
            }
            # ... other attributes

Required attributes are :attr:`~ModelFormOptions.model`, and either
:attr:`~ModelFormOptions.fields` or :attr:`~ModelFormOptions.exclude`. All
other ``Meta`` attributes are optional.

> Optional attributes, other than :attr:`~ModelFormOptions.localized_fields` and
> :attr:`~ModelFormOptions.formfield_callback`, expect a dictionary that maps a

Reviewer: ```suggestion
Most optional attributes, except :attr:`~ModelFormOptions.localized_fields` and
:attr:`~ModelFormOptions.formfield_callback`, expect a dictionary mapping a
```
(perhaps)

model field name to a value. Any field that is not defined in the dictionary
falls back to the field's default value.

.. admonition:: Invalid field names

>     An invalid or excluded field name strings in an optional dictionary
>     attribute will have no effect since fields that are not included are not
>     accessed.

Reviewer: ```suggestion
    Invalid or excluded field names in an optional dictionary attribute have no
    effect, since fields that are not included are not accessed.
```
(perhaps)

.. admonition:: Invalid Meta class attributes

>     You may place any attributes on a ``Meta`` class, but **only those listed
>     above are recognized**. If behavior seems unexpected, check for typos or
>     incorrect attribute names. Such mistakes **won’t themselves raise an
>     error** (unless they indirectly cause a required attribute to be wrong or
>     missing).

Reviewer: ```suggestion
    You may define any attribute on a ``Meta`` class, and hence typos or
    incorrect attribute names do not raise an error.
```
(perhaps)

``error_messages``
------------------

.. attribute:: ModelFormOptions.error_messages

>     A dictionary that maps a model field name to a dictionary of error message 
>     keys (``null``, ``blank``, ``invalid``, ``unique``, etc.) mapped to custom 
>     error messages.  
>     
>     Where a field doesn't have an error message dictionary mapped to it in
>     :attr:`~ModelFormOptions.error_messages`, Django will fall back on error
>     messages defined in that model field's
>     :attr:`django.db.models.Field.error_messages` and then finally on the
>     default error messages for that field type.

Reviewer: ```suggestion
    A dictionary that maps a model field name to a dictionary of error message
    keys (``null``, ``blank``, ``invalid``, ``unique``, etc.) mapped to custom
    error messages.

    When a field is not specified, Django will fall back on the error messages
    defined in that model field's :attr:`django.db.models.Field.error_messages`
    and then finally on the default error messages for that field type.
```

``exclude``
-----------

.. attribute:: ModelFormOptions.exclude

>     A tuple of strings or list of strings where each string is the name of a 
>     field from the model set in :attr:`~ModelFormOptions.model`. 
>     One of :attr:`~ModelFormOptions.fields` or :attr:`~ModelFormOptions.exclude` 
>     must be set.
>     Every field named in :attr:`~ModelFormOptions.exclude` will be excluded from the output form,
>     **including fields that are also listed in** :attr:`~ModelFormOptions.fields`.
> 
>     * When using :attr:`~ModelFormOptions.exclude`, leaving
>       :attr:`~ModelFormOptions.fields` unset or setting it to ``None`` has the
>       net effect of including all valid model fields (as if
>       :attr:`~ModelFormOptions.fields` were set to ``'__all__'``) except for
>       those provided to :attr:`~ModelFormOptions.exclude`. 
>     * If neither of :attr:`~ModelFormOptions.fields` or
>       :attr:`~ModelFormOptions.exclude` are set, an
>       :class:`~django.core.exceptions.ImproperlyConfigured` exception will be
>       raised.

Reviewer: ```suggestion
    A tuple or list of :attr:`~ModelFormOptions.model` field names to be
    excluded from the form.

    Either :attr:`~ModelFormOptions.fields` or
    :attr:`~ModelFormOptions.exclude` must be set. If neither are set, an
    :class:`~django.core.exceptions.ImproperlyConfigured` exception will be
    raised. If :attr:`~ModelFormOptions.exclude` is set and
    :attr:`~ModelFormOptions.fields` is unset, all model fields, except for
    those specified in :attr:`~ModelFormOptions.exclude`, are included in the
    form.
```

``field_classes``
-----------------

.. attribute:: ModelFormOptions.field_classes

>     A dictionary that maps a model field name string to a
>     :class:`~django.forms.Field` class (**not** a string). 
> 
>     :attr:`~ModelFormOptions.field_classes` can be used to set or
>     override the :class:`django.forms.Field` type for a field, including the
>     default form field associated with a given model :class:`django.db.models.Field` type.
> 
>     The :class:`~django.forms.Field` specified in
>     :attr:`~ModelFormOptions.field_classes` (the 'new field') gets
>     passed all the arguments that would otherwise have been used for that
>     model's field (the 'old field'). If the old field has different attributes
>     (and therefore different arguments) than the new field, a :exc:`TypeError`
>     may be raised. For example:
>     
>     * Replacing a :class:`~django.forms.CharField` with an
>       :class:`~django.forms.EmailField` will work, since the latter is a
>       subclass of the former and can handle all the arguments passed to it.
>     * Replacing a :class:`~django.forms.CharField` with an
>       :class:`~django.forms.IntegerField`, however, will **not** work, since the
>       :class:`~django.forms.CharField` would pass a
>       :attr:`~django.forms.CharField.max_length` argument, and the
>       :class:`~django.forms.IntegerField` cannot accept such an argument (it
>       instead has :attr:`~django.forms.IntegerField.max_value`). In this case, a
>       :exc:`TypeError` will be raised with the message: 
>     
>         .. code-block:: console
> 
>             TypeError: Field.__init__() got an unexpected keyword argument 'max_length'

Reviewer: ```suggestion
    A dictionary that maps a model field name to a :class:`~django.forms.Field`
    class, which overrides the ``form_class`` used in the model field's
    :meth:`.Field.formfield` method.

    When a field is not specified, Django will fall back on the model field's
    :ref:`default field class <model-form-field-types>`.
```
Note that I have added the reference:
```diff
--- a/docs/topics/forms/modelforms.txt
+++ b/docs/topics/forms/modelforms.txt
@@ -38,6 +38,8 @@ For example:
     >>> article = Article.objects.get(pk=1)
     >>> form = ArticleForm(instance=article)
 
+.. _model-form-field-types:
+
 Field types
 -----------
```

``fields``
----------

.. attribute:: ModelFormOptions.fields

>     One of :attr:`~ModelFormOptions.fields` or :attr:`~ModelFormOptions.exclude`
>     must be defined.
> 
>     Valid values:
> 
>     * The special string ``'__all__'`` — include every editable field on
>       :attr:`~ModelFormOptions.model` (non‑editable fields are always excluded).
>     * A tuple or list of distinct editable field name strings written in
>       snake_case (e.g. ``('title', 'author')`` or ``['title', 'author']``).
> 
>     Any other input (a lone string that is not ``'__all__'``, a value that
>     is not a tuple/list, inclusion of invalid or non‑editable field names, or
>     omitting both :attr:`~ModelFormOptions.fields` and
>     :attr:`~ModelFormOptions.exclude`) will cause Django to raise an appropriate
>     exception (e.g. :exc:`TypeError`, :exc:`~django.core.exceptions.FieldError`,
>     or :exc:`~django.core.exceptions.ImproperlyConfigured`).

Reviewer: ```suggestion
    A tuple or list of :attr:`~ModelFormOptions.model` field names to be
    included in the form. The value ``'__all__'`` can be used to specify that
    all fields should be included.

    If any field is specified in :attr:`~ModelFormOptions.exclude`, this will
    not be included in the form despite being specified in
    :attr:`~ModelFormOptions.fields`.

    Either :attr:`~ModelFormOptions.fields` or
    :attr:`~ModelFormOptions.exclude` must be set. If neither are set, an
    :class:`~django.core.exceptions.ImproperlyConfigured` exception will be
    raised.
```

``formfield_callback``
----------------------

.. attribute:: ModelFormOptions.formfield_callback

>     A function or callable that takes a model field (**not** a string,
>     though the name of the field can be accessed in the called function using
>     ``modelField.name``) and returns a :class:`django.forms.Field` object.
> 
>     If :attr:`~ModelFormOptions.formfield_callback` is neither a callable nor a
>     function, a :exc:`TypeError` will be raised. 

Reviewer: ```suggestion
    A function or callable that takes a model field and returns a
    :class:`django.forms.Field` object.
```

``help_texts``
--------------

.. attribute:: ModelFormOptions.help_texts

>     A dictionary that maps a model field name string to a help text
>     string. 
>     
>     Where a field doesn't have a label string mapped to it in
>     :attr:`~ModelFormOptions.help_texts`, Django will fall back on that model
>     field's :attr:`~django.db.models.Field.help_text`.

Reviewer: ```suggestion
    A dictionary that maps a model field name to a help text string.

    When a field is not specified, Django will fall back on that model field's
    :attr:`~django.db.models.Field.help_text`.
```

``labels``
----------

.. attribute:: ModelFormOptions.labels

>     A dictionary that maps a model field name string to a label string. 
>     
>     Where a field doesn't have a label string mapped to it in
>     :attr:`~ModelFormOptions.labels`, Django will fall back on that model
>     field's :attr:`~django.db.models.Field.verbose_name` and then the field's
>     attribute name.

Reviewer: ```suggestion
    A dictionary that maps a model field names to a label string.

    When a field is not specified, Django will fall back on that model field's
    :attr:`~django.db.models.Field.verbose_name` and then the field's attribute
    name.
```

``localized_fields``
--------------------

.. attribute:: ModelFormOptions.localized_fields

>     A tuple of strings or a list of strings that are field names.
>     
>     The fields included in this list will have their data localized (by default,
>     the form fields output by a :class:`~django.forms.ModelForm` are **not**
>     localized, see :ref:`Creating forms from models: Enabling localization of
>     fields <modelforms-enabling-localization-of-fields>`).  
>     
>     Like :attr:`~ModelFormOptions.fields`,
>     :attr:`~ModelFormOptions.localized_fields` can be set to the special string
>     ``'__all__'`` to include all fields.

Reviewer: ```suggestion
    A tuple or list of :attr:`~ModelFormOptions.model` field names to be
    localized. The value ``'__all__'`` can be used to specify that all fields
    should be localized.

    By default, form fields are not localized, see
    :ref:`enabling localization of fields
    <modelforms-enabling-localization-of-fields>` for more details.
```

``model``
---------

.. attribute:: ModelFormOptions.model

>     *Required.* A :class:`django.db.models.Model` to be used for this
>     :class:`~django.forms.ModelForm`.
> 
>     If no class is specified, a :exc:`ValueError` exception will be raised.

Reviewer: ```suggestion
    Required. The :class:`django.db.models.Model` to be used for the
    :class:`~django.forms.ModelForm`.
```

``widgets``
-----------

.. attribute:: ModelFormOptions.widgets

>     A dictionary that maps a model field name string to a
>     :class:`django.forms.Widget`. 
>     
>     Where an included model field doesn't have a widget mapped to it in
>     :attr:`~ModelFormOptions.widgets`, Django will fall back on the default
>     widget for that particular type of :class:`django.db.models.Field`.

Reviewer: ```suggestion
    A dictionary that maps a model field name to a
    :class:`django.forms.Widget`.

    When a field is not specified, Django will fall back on the default widget
    for that particular type of :class:`django.db.models.Field`.
```
